Updates

dwillcox opened the pull request:
- Fixed an problem where integrator.f90.template assumed free protons in the reaction network.
 
- Generalized prefactor and dens_exp calculation.

- Added triple-alpha and triple-alpha_f90 examples.
 
- Added a test-chapters example which uses one rate from each reaclib chapter, as a start for unit testing.

zingale commented:
I just pushed 2 fixes to the original source:

-- got the notebook working again (it was broken from the last PR) + it now works with python3
-- for the `examples/CNO/`, there is now a file `cno_rhs.py.benchmark` -- this is the "correct" output. 

Can you confirm that your changes here still give the correct output for CNO?  Just run `cno.py` and diff with the benchmark file

Can you add benchmark files for your nets too?  This will be the start of the unit testing -- we'll ensure that we always get the right answer for all subsequent pulls.

zingale commented:
and yes, I should have done my changes as a PR too :(

dwillcox commented:
My version does give the correct output, though not the same output. Mine reverses the order of the Y[in]*Y[im] multiplications to read Y[im]*Y[in].

I've added unit testing with nose, which the readme in pyreaclib/test explains. Because I had been working on this before today, I didn't use the .benchmark scheme. Instead, there is a directory pyreaclib/test/standard with the same subdirectory names as the examples directories. Each directory pyreaclib/test/standard/[example] contains the expected output from its corresponding example script, which prevents the pyreaclib/examples directories from becoming cumbersome, especially with the fortran examples which produce a lot of output files.

pyreaclib/test/test_reaclib.py runs the nose unit testing as follows:

1) Create a directory pyreaclib/test/runs/XYZ where XYZ is formed from the current date and time

2) Copy the contents of pyreaclib/examples/[example] to pyreaclib/test/runs/XYZ/[example] for each example

3) Run the example scripts in pyreaclib/test/runs/XYZ

4) Use diff to compare each file in pyreaclib/test/standard/[example] with its corresponding file in pyreaclib/test/runs/XYZ/[example], asserting the diff output is an empty string.

The contents of pyreaclib/test/runs will not be tracked by git.

Also, the 'chapters' example I added containing 11 nuclides and 11 rates, 1 rate from each Reaclib chapter, I have added to the unit testing after manually verifying the following about chapters_rhs.py:

1) All the reaclib fit parameters are read and written correctly

2) All the prefactors, exponents, etc in all the RHS terms are correct.

dwillcox commented:
I just merged the upstream changes into this pull request and removed the .benchmark file. When I ran cno.py in a repo I cloned directly from pyreaclib/pyreaclib.git, I also see that the order of multiplication doesn't match what's in cno_rhs.py.benchmark so I suspect it came from an earlier git commit than the current one in pyreaclib/pyreaclib.git...

dwillcox commented:
I was looking further into the benchmark file and it didn't come from 949bc17ffce5cc02fa799c0356e95ff233a1f22b either, which was the last commit before we merged in all my stuff. Running diff on cno_rhs.py between that commit and the latest in the current pull request, they match.

dwillcox commented:
Fixed a formatting issue with the readme in pyreaclib/test
